### ISSUE TITLE
Techempower Benchmark

### ISSUE THREAD

**Issue by szmarczak** [uNetworking/uWebSockets#835]
How does this compare to the [Techempower Benchmark](https://www.techempower.com/benchmarks/)? It's kinda hard to believe the benchmarks you have provided - the source code is incomplete (it lacks code for other HTTP libs).

**Comment by alexhultman** [comment]
Believe you can do in church. Here we only deal with scientific measurements of which you can do yourself (that's the entire point of science, reproducibility).

> benchmarks you have provided - the source code is incomplete (it lacks code for other HTTP libs).

Please explain, I don't get what you're trying to say. There's no source code specific to any one server, they all share the same exact benchmarking code - the standard wrk tool using default settings.

**Comment by szmarczak** [comment]
> Please explain, I don't get what you're trying to say.

I mean that it lacks server code for different projects like Deno, Express, Node etc.
Or is it just simplest `Hello, world` ?

> the standard wrk tool using default settings

Have tried out [autocannon](https://github.com/mcollina/autocannon)? For me it provides more load than `wrk` (about 20k+ more requests when testing Node).

Looking back at my first question, I'm really curious how `uws` compares to the Techempower Benchmark.
It would be nice to see it there.

**Comment by alexhultman** [comment]
This project is not a database or a json serializer.

So it makes little sense to benchmark it as if. Besides, TechEmpower cannot properly benchmark high end server software, the top 13 are all separated by nothing but noise. They are all capped by IO.

Further on TechEmpower do not properly normalize CPU-time, leading to majorly invalid results where single threaded servers are compared with multithreaded ones, etc.

In short TechEmpower is really not a good benchmark. I do my own tests and if they are not enough then do your own tests.

This project is not in a phase where it needs to prove anything regarding performance. It's pretty much a well known fact by now.

I have an upcoming article benchmarking 1 million websockets on a laptop over WiFi. That's the kind of stuff I do.

**Comment by ddevienne** [comment]
If you don't mind, could you please post your article in this thread or your main project page?
Also, for reference, a recent "1 million websockets" in Go: https://github.com/eranyanay/1m-go-websockets
I'm eager to read your article!

**Comment by szmarczak** [comment]
> This project is not a database or a json serializer. So it makes little sense to benchmark it as if.

It doesn't have to parcitipate in all benchmark categories. It can just take part in `/plaintext`.

> top 13 are all separated by nothing but noise

It's fine till you need to handle more users, you need a much better machine - then yes, it will have a difference, obviously.

> Further on TechEmpower do not properly normalize CPU-time, leading to majorly invalid results where single threaded servers are compared with multithreaded ones, etc.

Have you [checked](https://github.com/TechEmpower/FrameworkBenchmarks) that? Don't blame them if you haven't. Also, they use `wrk` too.

> This project is not in a phase where it needs to prove anything regarding performance. It's pretty much a well known fact by now.

Well, you haven't convinced me. Because you say so it doesn't mean it is.

**Comment by alexhultman** [comment]
I don't need to convince you. You can pay me xxx USD an hour and I will try my hardest. Otherwise, not interested.

**Comment by alexhultman** [comment]
Published here https://medium.com/@alexhultman/millions-of-active-websockets-with-node-js-7dc575746a01

**Comment by szmarczak** [comment]
Why did you use `--max-old-space-size=8192` while the app used only 0.5GB? `1024` should be enough in that case.

**Comment by alexhultman** [comment]
The point being made is you can't do the same with other libs. That process was well over 4 GB
